src/main/java/com/multithreading/TextProcessor.java

```diff
@@ -1,38 +1,67 @@
 package com.multithreading;
 
+import java.util.ArrayList;
+import java.util.List;
+
 class Capitalize implements Runnable {
-    private final char[] chars;
-    private final int index;
+    private volatile boolean isRunning = true;
+    private char character;
 
-    public Capitalize(char[] chars, int index) {
-        this.chars = chars;
-        this.index = index;
+    public Capitalize(char character) {
+        this.character = character;
     }
 
     @Override
     public void run() {
-        synchronized (chars) {
-            chars[index] = Character.toUpperCase(chars[index]);
+        if (isRunning) {
+            this.character = Character.toUpperCase(character);
         }
     }
+
+    public void stop() {
+        this.isRunning = false;
+    }
+
+    public char getChar() {
+        return character;
+    }
 }
 
 public class TextProcessor {
 
+    private final List<Thread> threads;
+    private final List<Capitalize> capitalizeTasks;
+
+    TextProcessor() {
+        this.capitalizeTasks = new ArrayList<>();
+        this.threads = new ArrayList<>();
+    }
+
     public String process(String input) throws InterruptedException {
-        Thread[] threads = new Thread[input.length()];
+        int inputLength = input.length();
 
         char[] chars = input.toCharArray();
+        char[] capitalizedChars = new char[inputLength];
 
-        for (int i = 0; i < chars.length; i++) {
-            threads[i] = new Thread(new Capitalize(chars, i));
-            threads[i].start();
+        for (int i = 0; i < inputLength; i++) {
+            capitalizeTasks.add(new Capitalize(chars[i]));
+            threads.add(new Thread(capitalizeTasks.getLast()));
+            threads.getLast().start();
         }
 
         for (Thread thread : threads) {
             thread.join();
```

Comment:
how about you catch interrupted exception here but still try to join all the threads. And after the for loop you clean-up and throw interrupted if any thread has thrown interrupted, otherwise you return the capitalized String from joining the result from all the tasks?

Reply:
Another option instead of joining the threads (which means your processor knows about the threads vs knows about the tasks), is to keep track of all the tasks, then wait for all the tasks to complete (within X amount of time). If the tasks are not completed by X, then shutdown the non-completed tasks and throw interrupted.

2 methods: push vs pull
(1) push
you have a state boolean on each of those tasks. While loop to check which tasks are completed + sleep.

(2) pull
pass a callback to the task to notify parent when the taskId is completed. While !all completed sleep().

Either case, you ignore the thread itself. Let the thread die when the task runnable die.